File: ios/ios_new/ACLRehabTracker/Services/FirestoreService.swift
import Foundation
import FirebaseFirestore

class FirestoreService {
    static let shared = FirestoreService()
    private let db = Firestore.firestore()

    private init() {}

    // MARK: - User Profile

    /// Save user profile to Firestore
    func saveUserProfile(uid: String, profile: UserProfile) async throws {
        let profileRef = db.collection("users").document(uid).collection("profile").document("info")
        try profileRef.setData(from: profile)
    }

    /// Get user profile from Firestore
    func getUserProfile(uid: String) async throws -> UserProfile? {
        let profileRef = db.collection("users").document(uid).collection("profile").document("info")
        let document = try await profileRef.getDocument()

        guard document.exists else { return nil }

        var profile = try document.data(as: UserProfile.self)
        profile.id = document.documentID
        return profile
    }

    // MARK: - Measurements

    /// Save a measurement to Firestore
    func saveMeasurement(uid: String, measurement: Measurement) async throws -> String {
        let measurementsRef = db.collection("users").document(uid).collection("measurements")
        let docRef = try measurementsRef.addDocument(from: measurement)
        return docRef.documentID
    }

    /// Get all measurements for a user, ordered by timestamp descending
    func getMeasurements(uid: String) async throws -> [Measurement] {
        let measurementsRef = db.collection("users").document(uid).collection("measurements")
        let query = measurementsRef.order(by: "timestamp", descending: true)
        let snapshot = try await query.getDocuments()

        return snapshot.documents.compactMap { document in
            var measurement = try? document.data(as: Measurement.self)
            measurement?.id = document.documentID
            return measurement
        }
    }

    /// Get measurements filtered by type
    func getMeasurementsByType(uid: String, type: MeasurementType) async throws -> [Measurement] {
        let measurementsRef = db.collection("users").document(uid).collection("measurements")
        let query = measurementsRef
            .whereField("type", isEqualTo: type.rawValue)
            .order(by: "timestamp", descending: true)
        let snapshot = try await query.getDocuments()

        return snapshot.documents.compactMap { document in
            var measurement = try? document.data(as: Measurement.self)
            measurement?.id = document.documentID
            return measurement
        }
    }

    /// Delete a measurement
    func deleteMeasurement(uid: String, measurementId: String) async throws {
        let measurementRef = db.collection("users").document(uid).collection("measurements").document(measurementId)
        try await measurementRef.delete()
    }

    /// Update photo URL for a measurement
    func updateMeasurementPhotoUrl(uid: String, measurementId: String, photoUrl: String) async throws {
        let measurementRef = db.collection("users").document(uid).collection("measurements").document(measurementId)
        try await measurementRef.updateData(["photoUrl": photoUrl])
    }

    /// Delete all user data (profile + measurements)
    func deleteUserData(uid: String) async throws {
        // Delete profile
        let profileRef = db.collection("users").document(uid).collection("profile").document("info")
        try await profileRef.delete()

        // Delete all measurements
        let measurementsRef = db.collection("users").document(uid).collection("measurements")
        let snapshot = try await measurementsRef.getDocuments()
        for document in snapshot.documents {
            try await document.reference.delete()
        }
    }
}

File: android/app/src/main/java/com/twintipsolutions/aclrehabtracker/data/service/FirestoreService.kt
package com.twintipsolutions.aclrehabtracker.data.service

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.twintipsolutions.aclrehabtracker.data.model.Measurement
import com.twintipsolutions.aclrehabtracker.data.model.MeasurementType
import com.twintipsolutions.aclrehabtracker.data.model.UserProfile
import kotlinx.coroutines.tasks.await

object FirestoreService {
    private val db = FirebaseFirestore.getInstance()

    // --- User Profile ---

    suspend fun saveUserProfile(uid: String, profile: UserProfile) {
        db.collection("users").document(uid)
            .collection("profile").document("info")
            .set(profile.toFirestore())
            .await()
    }

    suspend fun getUserProfile(uid: String): UserProfile? {
        val doc = db.collection("users").document(uid)
            .collection("profile").document("info")
            .get()
            .await()
        val data = doc.data ?: return null
        return UserProfile.fromFirestore(uid, data)
    }

    // --- Measurements ---

    suspend fun saveMeasurement(uid: String, measurement: Measurement): String {
        val docRef = db.collection("users").document(uid)
            .collection("measurements")
            .add(measurement.toFirestore())
            .await()
        return docRef.id
    }

    suspend fun getMeasurements(uid: String): List<Measurement> {
        val snapshot = db.collection("users").document(uid)
            .collection("measurements")
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            doc.data?.let { Measurement.fromFirestore(doc.id, it) }
        }
    }

    suspend fun getMeasurementsByType(uid: String, type: MeasurementType): List<Measurement> {
        val snapshot = db.collection("users").document(uid)
            .collection("measurements")
            .whereEqualTo("type", type.name.lowercase())
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .get()
            .await()
        return snapshot.documents.mapNotNull { doc ->
            doc.data?.let { Measurement.fromFirestore(doc.id, it) }
        }
    }

    suspend fun updateMeasurementPhotoUrl(uid: String, measurementId: String, photoUrl: String) {
        db.collection("users").document(uid)
            .collection("measurements").document(measurementId)
            .update("photoUrl", photoUrl)
            .await()
    }

    suspend fun deleteMeasurement(uid: String, measurementId: String) {
        db.collection("users").document(uid)
            .collection("measurements").document(measurementId)
            .delete()
            .await()
    }

    suspend fun deleteUserData(uid: String) {
        val batch = db.batch()

        // Delete profile
        val profileRef = db.collection("users").document(uid)
            .collection("profile").document("info")
        batch.delete(profileRef)

        // Delete all measurements
        val snapshot = db.collection("users").document(uid)
            .collection("measurements")
            .get()
            .await()
        for (doc in snapshot.documents) {
            batch.delete(doc.reference)
        }

        batch.commit().await()
    }
}
